### DISCUSSION TITLE
How to fetch data from server once then make it global for every page

### DISCUSSION THREAD

**Discussion by bymoe** [vercel/next.js#14913] [Help] [answered]
I have global data that I need to get from server once, then make it global for every page, As simple as this.

```
const MyApp = ({ Component, pageProps, globalLayout }) => {
  return <Component {...pageProps} globalLayout={globalLayout} />;
};

MyApp.getInitialProps = async ({ req }) => {
  if (typeof window === "undefined") {
    const subdomain = req.headers.host.split(".")[0];
    const res = await fetch(
      `https://myapi.com/api/v1/layout?sub_domain=${subdomain}`
    );
    const data = await res.json();

    return {
      globalLayout: data.results ? data.results.portal : null,
    };
  }
};

export default MyApp;
```

Data depends on the subdomain, however I am getting `TypeError: Cannot read property 'headers' of undefined`

I was thinking about using `getServerSideProps` in `_app.js` and pass it to every component, but It's not supported yet. Any ideas?

**Comment by srikanth912** [comment, marked as answer]
I believe in _app.js , req is part of ctx so you need to do something like below: 

async ({ ctx: { req } }) => {
  console.log(req.host);
};

**Reply by bymoe** [reply]
How did I miss this? Really thank you!

**Reply by rubenzup** [reply]
my current _app.tsx is like this, how could i do one fetch and pass to any component i want? 


`function MyApp({ Component, pageProps }: AppProps) {
    return <Component {...pageProps} />;
}`

**Reply by luigi-derson** [reply]
Actually if you come from the docs you would have seen that `getInitalProps` only passes one argument, that is right for all pages but `_app.js` file. In this case you also have access to `Component` argument. `getInitialProps({ Component, ctx })`
